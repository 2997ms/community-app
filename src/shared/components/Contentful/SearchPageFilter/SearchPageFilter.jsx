/**
 * The core search bar rendering.
 */
import _ from 'lodash';
import moment from 'moment';
import PT from 'prop-types';
import React, { Component } from 'react';
import { themr } from 'react-css-super-themr';

import IconCloseBig from 'assets/images/tc-edu/icon-close-big.svg';
import IconArrowUpSmall from 'assets/images/tc-edu/icon-arrow-up-small.svg';
import FilterTags from './FilterTags';
import FilterAuthor from './FilterAuthor';
import FilterDate from './FilterDate';
import FilterRadio from './FilterRadio';
import FilterSelection from './FilterSelection';
import defaultTheme from './themes/default.scss';


export class SearchPageFilterInner extends Component {
  constructor(props) {
    super(props);

    const { categories, selectedCategory } = props;
    const selectedCategoryObject = _.find(categories, { name: selectedCategory });
    this.state = {
      isShowSubCategory: true,
      selectedAuthor: props.selectedAuthor,
      authorList: props.authorList,
      tags: _.cloneDeep(props.tags),
      startDate: props.startDate,
      endDate: props.endDate,
      selectedCategory: selectedCategoryObject ? _.cloneDeep(selectedCategoryObject) : null,
      categories,
    };
  }

  render() {
    const {
      theme,
      onClose,
      className,
      isShowInMobile,
    } = this.props;

    const {
      isShowSubCategory,
      selectedAuthor,
      authorList,
      startDate,
      endDate,
      tags,
      selectedCategory,
      categories,
    } = this.state;

    return (
      <div className={`${theme.container} ${className} ${isShowInMobile ? '' : theme['is-desktop']}`}>
        <div className={`${theme.header} ${theme['is-mobile']}`}>
          <span>filter</span>
          <button
            type="button"
            onClick={onClose}
          >
            <IconCloseBig />
          </button>
        </div>
        <div className={theme.section}>
          <div
            className={theme['section-header']}
          >
            <span className={theme['section-title']}>category</span>
          </div>
          <FilterRadio
            selected={selectedCategory ? selectedCategory.name : null}
            options={categories}
            onSelected={(item) => {
              if (item) {
                const newSelectedCategory = _.cloneDeep(item);
                for (let i = 0; i < newSelectedCategory.subCategories.length; i += 1) {
                  newSelectedCategory.subCategories[i].selected = false;
                }
                this.setState({ selectedCategory: newSelectedCategory });
              }
            }}
          />
        </div>
        <div className={theme.section}>
          <button
            type="button"
            className={`${theme['section-header']} ${isShowSubCategory ? theme.expanded : ''}`}
            onClick={() => { this.setState({ isShowSubCategory: !isShowSubCategory }); }}
          >
            <span className={theme['section-title']}>sub category</span>
            <IconArrowUpSmall />
          </button>
          {selectedCategory && isShowSubCategory && (
            <FilterSelection
              options={selectedCategory.subCategories}
              onSelected={(index) => {
                const subCategory = selectedCategory.subCategories[index];
                subCategory.selected = !subCategory.selected;
                this.setState({ selectedCategory });
              }}
            />
          )}
        </div>
        <div className={theme.section}>
          <div
            type="button"
            className={theme['section-header']}
          >
            <span className={theme['section-title']}>filter</span>
          </div>
          <FilterTags
            tags={tags}
            onAddNewTag={(tag) => {
              tags.push(tag);
              this.setState({ tags });
            }}
            onRemoveTag={(index) => {
              tags.splice(index, 1);
              this.setState({ tags });
            }}
          />
          <FilterAuthor
            className={theme['author-container']}
            selected={selectedAuthor}
            options={authorList}
            onSelected={(item) => { this.setState({ selectedAuthor: item }); }}
          />
          <FilterDate
            className={theme['track-date-container']}
            startDate={startDate}
            endDate={endDate}
            onSelectStartDate={(date) => { this.setState({ startDate: date }); }}
            onSelectEndDate={(date) => { this.setState({ endDate: date }); }}
          />
        </div>
        <div className={theme.bottom}>
          <button
            type="button"
            className={theme['btn-apply']}
          >APPLY FILTER
          </button>
        </div>
      </div>
    );
  }
}

SearchPageFilterInner.defaultProps = {
  onClose: () => {},
  selectedAuthor: '',
  authorList: [],
  startDate: null,
  endDate: null,
  tags: [],
  selectedCategory: '',
  categories: [],
  className: '',
  isShowInMobile: false,
};

SearchPageFilterInner.propTypes = {
  theme: PT.shape({
    container: PT.string.isRequired,
    bottom: PT.string.isRequired,
    header: PT.string.isRequired,
    section: PT.string.isRequired,
    expanded: PT.string.isRequired,
    'track-date-container': PT.string.isRequired,
    'author-container': PT.string.isRequired,
    'btn-apply': PT.string.isRequired,
    'section-header': PT.string.isRequired,
    'section-title': PT.string.isRequired,
    'is-mobile': PT.string.isRequired,
    'is-desktop': PT.string.isRequired,
  }).isRequired,
  onClose: PT.func,
  selectedAuthor: PT.string,
  startDate: PT.instanceOf(moment),
  endDate: PT.instanceOf(moment),
  authorList: PT.arrayOf(PT.string),
  tags: PT.arrayOf(PT.string),
  selectedCategory: PT.string,
  categories: PT.arrayOf(PT.shape({
    name: PT.string.isRequired,
    subCategories: PT.arrayOf(PT.shape({
      name: PT.string.isRequired,
      selected: PT.bool,
    })),
  })),
  className: PT.string,
  isShowInMobile: PT.string,
};

export default themr('Contentful-Blog', defaultTheme)(SearchPageFilterInner);
