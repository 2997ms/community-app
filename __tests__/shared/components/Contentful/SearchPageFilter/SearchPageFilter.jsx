import React from 'react';
import Renderer from 'react-test-renderer/shallow';
import { SearchPageFilterInner } from 'components/Contentful/SearchPageFilter/SearchPageFilter';

test('Matches shallow shapshot', () => {
  const renderer = new Renderer();
  const MOCK_PROPS = {
    theme: {},
  };
  renderer.render(<SearchPageFilterInner {...MOCK_PROPS} />);
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});
