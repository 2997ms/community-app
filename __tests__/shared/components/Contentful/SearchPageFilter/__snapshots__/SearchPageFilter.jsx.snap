// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches shallow shapshot 1`] = `
<div
  className="undefined  undefined"
>
  <div
    className="undefined undefined"
  >
    <span>
      filter
    </span>
    <button
      onClick={[Function]}
      type="button"
    >
      <IconCloseBig
        height="14"
        viewBox="0 0 14 14"
        width="14"
        xmlns="http://www.w3.org/2000/svg"
      />
    </button>
  </div>
  <div>
    <div>
      <span>
        category
      </span>
    </div>
    <ThemedFilterRadioInner
      className=""
      composeAdhocTheme="deeply"
      composeContextTheme="softly"
      mapThemrProps={[Function]}
      onSelected={[Function]}
      options={Array []}
      selected={null}
      themePriority="adhoc-context-default"
    />
  </div>
  <div>
    <button
      className="undefined undefined"
      onClick={[Function]}
      type="button"
    >
      <span>
        sub category
      </span>
      <IconArrowUpSmall
        height="8"
        viewBox="0 0 12 8"
        width="12"
        xmlns="http://www.w3.org/2000/svg"
      />
    </button>
  </div>
  <div>
    <div
      type="button"
    >
      <span>
        filter
      </span>
    </div>
    <ThemedFilterTagsInner
      composeAdhocTheme="deeply"
      composeContextTheme="softly"
      mapThemrProps={[Function]}
      onAddNewTag={[Function]}
      onRemoveTag={[Function]}
      tags={Array []}
      themePriority="adhoc-context-default"
    />
    <ThemedFilterAuthorInner
      className=""
      composeAdhocTheme="deeply"
      composeContextTheme="softly"
      mapThemrProps={[Function]}
      onSelected={[Function]}
      options={Array []}
      selected=""
      themePriority="adhoc-context-default"
    />
    <ThemedFilterDateInner
      className=""
      composeAdhocTheme="deeply"
      composeContextTheme="softly"
      endDate={null}
      mapThemrProps={[Function]}
      onSelectEndDate={[Function]}
      onSelectStartDate={[Function]}
      startDate={null}
      themePriority="adhoc-context-default"
    />
  </div>
  <div>
    <button
      type="button"
    >
      APPLY FILTER
    </button>
  </div>
</div>
`;
